src/sparse/sparse_matrix.h

```diff
@@ -0,0 +1,131 @@
+/*!
+ *  Copyright (c) 2022 by Contributors
+ * \file sparse/sparse_matrix.h
+ * \brief DGL C++ sparse matrix header
+ */
+#ifndef DGL_SPARSE_SPARSE_MATRIX_H_
+#define DGL_SPARSE_SPARSE_MATRIX_H_
+
+#include <torch/custom_class.h>
+#include <torch/script.h>
+
+#include <memory>
+
+namespace dgl {
+namespace sparse {
+
+/*! \brief SparseFormat enumeration */
+enum SparseFormat { kCOO, kCSR, kCSC };
+
+/*! \brief CSR sparse structure */
+struct CSR {
+  torch::Tensor indptr, indices;
+  // The element order of the sparse format. In the SparseMatrix, we have data
+  // (value_) for each non-zero value. The order of non-zero values in (value_)
+  // may differ from the order of non-zero entries in the sparse format. So we
+  // store `e_order` in a sparse format to indicate its relative element order
+  // to the SparseMatrix. With `e_order`, we can retrieve the correct data for a
+  // sparse format, i.e., `value_[e_order]`. If `e_order` is not defined, this
+  // sparse format follows the same non-zero value order as the SparseMatrix.
+  torch::optional<torch::Tensor> e_order;
+};
+
+/*! \brief COO sparse structure */
+struct COO {
+  torch::Tensor row, col;
+  // The element order of the sparse format. In the SparseMatrix, we have data
+  // (value_) for each non-zero value. The order of non-zero values in (value_)
+  // may differ from the order of non-zero entries in the sparse format. So we
+  // store `e_order` in a sparse format to indicate its relative element order
+  // to the SparseMatrix. With `e_order`, we can retrieve the correct data for a
+  // sparse format, i.e., `value_[e_order]`. If `e_order` is not defined, this
+  // sparse format follows the same non-zero value order as the SparseMatrix.
+  torch::optional<torch::Tensor> e_order;
```

Comment:
Same here.

Reply:
Actually, COO does not need this because DGL's default is COO's (row,col) order is the same as the non-zero value order.

Reply:
Same here. Please take a look at the structure definition in 
 DGL https://github.com/dmlc/dgl/blob/master/include/dgl/aten/coo.h . Some of the members/flags should be useful.

Reply:
I add a TODO here. We may discuss and add those flags later.